### PR TITLE
Implementation of Basic Functionalities to Play Through a Quest

### PR DESCRIPTION
**Pull request by philipchenaux**
This is a further enhancement of the previous pull request on this branch that initially created the main structs for a quest. This pull request involves creating and playing through a quest. We have changed the following files:
```
include/quests/quest_state.h
src/quests/quests_state.c
tests/quests/test_quests_state.c
```
In this pull request we have included the following main functions:
```C
int add_achievement_to_quest(quest_t *quest, achievement_t *achievement_to_add);
int start_quest(quest_t *quest);
int complete_achievement(quest_t *quest, item_t *item_collected, npc_t *npc_met);
int is_quest_completed(quest_t *quest);
int quest_status(quest_t *quest);
item_t *quest_completed(quest_t *quest);
```
These are mostly still relatively basic function that allow for users to create a very simply quest.

We have also added the demo file for our presentation.

### CODE REVIEW

File: include/quests/quests_state.h [hunk 1]
```diff
@@ -115,4 +120,76 @@ int achievement_free(achievement_t *achievement);
...
+ */
+int quest_status(quest_t *quest);
+
+/* Rewards the prize to the player once the quest has been completed
+ * This also checks for completion, and does nothing if not completed
+ *
+ * Parameter:
+ * - quest: pointer to quest_t
+ * 
+ * Returns:
```
Comment (shayna-k) [inline include/quests/quests_state.h:187]: Just a tiny style thing but maybe you could keep the format of the 'returns' part of the comment consistent (like the previous function description); maybe something like:
- item_t / NULL* : ...

Comment (philipchenaux) [reply include/quests/quests_state.h:187]: ok thanks!

File: include/quests/quests_state.h [hunk 2]
```diff
@@ -115,4 +120,78 @@ int achievement_free(achievement_t *achievement);
...
+ * Parameter:
+ * - quest: pointer to quest_t to be started
+ * 
+ * Returns:
+ * - SUCCESS 
+ * - FAILURE
+ */
+int start_quest(quest_t *quest);
+
+/* completes an achievement in a quest by checking if the item and npc matches
```
Comment (firatciftci) [inline include/quests/quests_state.h:146]: Very minor thing: Capital "c" to begin the function header.

File: include/quests/quests_state.h [hunk 3]
```diff
@@ -115,4 +120,78 @@ int achievement_free(achievement_t *achievement);
...
+ * - item_collected: the item that has been collected
+ * - npc_met: the npc that was met
+ *
+ * Returns:
+ * - SUCCESS
+ * - FAILURE
+ */
+int complete_achievement(quest_t *quest, item_t *item_collected, npc_t *npc_met);
+
+/* checks if the quest is completed
```
Comment (firatciftci) [inline include/quests/quests_state.h:159]: Very minor thing: Capital "c" to begin the function header.

File: include/quests/quests_state.h [hunk 4]
```diff
@@ -115,4 +120,78 @@ int achievement_free(achievement_t *achievement);
...
+ *
+ * Parameter:
+ * - quest: pointer to quest_t
+ * 
+ * Returns: status code for various status in the quest:
+ * 0: quest has not been started
+ * 1: quest has been started but not completed
+ * 2: quest has been completed
+ */
+int quest_status(quest_t *quest);
```
Comment (firatciftci) [inline include/quests/quests_state.h:179]: I think having the function name as `get_quest_status` would make more sense.

File: include/quests/quests_state.h [hunk 5]
```diff
@@ -115,4 +120,78 @@ int achievement_free(achievement_t *achievement);
...
+ * - quest: pointer to quest_t
+ * 
+ * Returns:
+ * - item_t: an item (possible to be added into the inventory of the player)
+ *           If the quest is not completed the function will return a null pointer
+ * 
+ * Note:
+ * The status of the quest should first be checked before this function is called
+ */
+item_t *quest_completed(quest_t *quest);
```
Comment (firatciftci) [inline include/quests/quests_state.h:194]: I think having the function name as something like `quest_is_completed` would make more sense.

Comment (firatciftci) [inline include/quests/quests_state.h:194]: I think having the function name as something like `complete_quest` would make more sense.

File: include/quests/quests_structs.h [hunk 6]
```diff
@@ -39,17 +39,22 @@ typedef struct achievement_llist {
...
 /* 
  * This is the hashable struct for a quest 
+ * Elements:
+ * quest_id: the id of the quest
+ * achievement_list: linked list struct holding a list of
+ *                   achievements that make up a quest
+ * reward: reward of the quest is an item
+ * status: 0: quest has not been started
+ *         1: quest has been started but not completed
+ *         2: quest has been completed
```
Comment (firatciftci) [inline include/quests/quests_structs.h:49]: Is there any mechanism where the quest is failed? If it doesn't seem to difficult for you guys given the time constraint, I think adding a "failed quest" parameter for `-1` would be beneficial to create a fully comprehensive quest status mechanism. If not, I'll definitely suggest creating a backlog issue for this idea.

Comment (philipchenaux) [reply include/quests/quests_structs.h:49]: ok! I will  implement this now!

Comment (philipchenaux) [reply include/quests/quests_structs.h:49]: I added another function that fails a quest if necessary!

**Comment by firatciftci** [comment]
Issue Score: ✔️++

Comments:
Amazing work!

**Pull request merged**